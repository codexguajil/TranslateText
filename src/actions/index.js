export const storeTranslation = (translation) => ({
  type: 'STORE_TRANS',
  translation
})

export const deleteTrans = (id) => ({
  type: 'DELETE_TRANS',
  id
})

export const postMessage = (message) => ({
  type: 'MESSAGE',
  message
})