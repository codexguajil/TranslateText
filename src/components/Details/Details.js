import React from 'react'
import PropTypes from 'prop-types'
import { Link, withRouther } from 'react-router-dom'

export const Details = (props) => {
 const {translatedText, original} = props
    return (
        <div>
          <div className='banner'>
          <Link to="/translations" className="header-a">
            Return to Translations
          </Link>
          <Link to="/" className="header-a">
              Translate to French.
          </Link>
        </div>
          <div className="details-field">
            <section className="left">
              <h1>Fran√ßais</h1>
              {translatedText}
            </section>
            <section className="right">
              <h1>English</h1>
              {original}
            </section>
          </div>
        </div>
    )
}

export default Details

Details.propTypes = {
  id: PropTypes.string,
  translatedText: PropTypes.string,
  original: PropTypes.string
}