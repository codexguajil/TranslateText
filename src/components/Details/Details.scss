.details-field {
  background: #FFF;
  color: black;
  display: flex;
  flex-direction: row;
  font-size: 32px;
  height: 600px;
  margin: auto;
  margin-top: 50px;
  width: 95%;
}

.left {
  background: plum;
  height: 500px;
  margin: auto;
  padding: 20px;
  overflow: scroll;
  width: 350px;
}

.right {
  background: cadetblue;
  height: 500px;
  margin: auto;
  padding: 20px;
  overflow: scroll;
  width: 350px;
}