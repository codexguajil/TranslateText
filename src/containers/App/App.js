import React, { Component } from 'react';
import { connect } from 'react-redux'
import { Route, withRouter } from 'react-router-dom'
import Form from '../../components/Form/Form';
import { storeTranslation } from '../../actions';
import PhraseContainer from '../PhraseContainer/PhraseContainer';
import key from '../../utils/apiKEY';
import PropTypes from 'prop-types';
import {Details} from '../../components/Details/Details';
import {postMessage} from '../../actions';

export class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoading: false,
    }
  }

  handleError = (message) => {
    this.props.postMessage(message)
    setTimeout(() => {
      this.props.postMessage('')
    }, 2000)
  }

  translateWords = async (content) => {
    this.setState({isLoading: true})
    const shortid = require('shortid');
    let newPhrase;
    const url = `https://translation.googleapis.com/language/translate/v2?key=${key}&format=text`
    try {
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(content),
      })
      newPhrase = await response.json()
    } catch(error) {
      return error.message
    }
    this.setState({isLoading: false})
    this.props.storeTranslation({...newPhrase.data.translations[0], id: shortid.generate(), original: content.q})
  }

  handleFormSubmit = (text, lang) => {
    let content = {
      'q': text,
      'target': lang
    }
    this.translateWords(content)
  }

  findTranslation = ({match}) => {
    const foundTranslation = this.props.translations.find(card => card.id === match.params.id)
    if(!foundTranslation) {
      return '404 no translation found!'
    }
    return <Details {...foundTranslation} />
  }

  render() {
    const {message} = this.props
    return (
      <div className="App">
       <Route exact path='/' component={ () => <Form handleSubmit={this.handleFormSubmit} handleError={this.handleError} translations={this.props.translations} />} />
       <p className='message'>{message}</p>
       <Route exact path='/translations' component={ () => <PhraseContainer loading={this.state.isLoading} /> } />
       <Route path='/translations/:id' render={this.findTranslation} />
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  translations: state.translations,
  message: state.message
})

export const mapDispatchToProps = (dispatch) => ({
  storeTranslation: (translations) => dispatch(storeTranslation(translations)),
  postMessage: (message) => dispatch(postMessage(message))
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));

App.propTypes = {
  translations: PropTypes.array,
  storeTranslation: PropTypes.func
}