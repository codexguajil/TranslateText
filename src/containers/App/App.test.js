import React from 'react';
import ReactDOM from 'react-dom';
import {App, mapStateToProps, mapDispatchToProps} from './App';
import {storeTranslation, postMessage} from '../../actions'
import {shallow} from 'enzyme'
import {Details} from '../../components/Details/Details'

jest.mock('../../actions')

describe('App', () => {
  let wrapper;

  it.skip('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should match the snapshot', () => {
    wrapper = shallow(
      <App />
    )

    expect(wrapper).toMatchSnapshot()
  })

  describe('mapStateToProps', () => {
    it('should return the translations in store', () => {
      const mockState = {
        translations: [],
        items: {}
      }

      const expected = {
        translations: []
      }

      const mockProps = mapStateToProps(mockState)
      expect(mockProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('calls dispatch with a storeTranslation action when storeTranslation is called', () => {
      const mockTranslation = {
        translatedText:"Bonjour.",
        detectedSourceLanguage:"en",
        id:"kVqASjuyE",
        original:"hello."
      }
      const mockDispatch = jest.fn()
      const actionToDispatch = storeTranslation(mockTranslation)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.storeTranslation(mockTranslation)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('calls dispatch with a message if there`s an error', () => {
      const mockError = 'this is an error'

      const mockDispatch = jest.fn()
      const actionToDispatch = postMessage(mockError)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.postMessage(mockError)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })

  describe('handleFormSubmit', () => {
    it('should call translateWords with the submitted text to translate', () => {
      wrapper = shallow(
        <App />
      )

      wrapper.instance().translateWords = jest.fn()

      let mockText = 'hello;'
      let expected = {
        'q': mockText,
        'target': 'fr'
      }

      wrapper.instance().handleFormSubmit(mockText)
      expect(wrapper.instance().translateWords).toHaveBeenCalledWith(expected)
    })
  })

  describe('translateWords', () => {
    it('should update the store with a translation object', async () => {
      let props = {
        storeTranslation: jest.fn()
      }

      wrapper = shallow(
        <App {...props} />
      )

      const mockContent = {
        'q': 'hello',
        'target': 'fr'
      }

      const mockTranslation = {
        translatedText:"Bonjour.",
        detectedSourceLanguage:"en",
        id:"kVqASjuyE",
        original:"hello."
      }

       await wrapper.instance().translateWords(mockContent)
       await expect(wrapper.instance().props.storeTranslation).toHaveBeenCalled()
    })

    it('should return an error if missing url or object', async () => {
      window.fetch = jest.fn(() => { return Promise.reject('error')})
        try {
          await fetch()
        } catch (error) {
          expect(error).toEqual('error')
        }
    })

    it('should set state when translateWords is called', async () => {
      wrapper = shallow(
        <App />
        )
        
      wrapper.instance().setState = jest.fn()
      await wrapper.instance().translateWords()
      await expect(wrapper.instance().setState).toHaveBeenCalled()
    })

    it('should set state to true when translateWords runs', async () => {
      wrapper = shallow(
        <App />
      )

      expect(wrapper.state('isLoading')).toEqual(false)
      await wrapper.instance().translateWords('hello')
      await expect(wrapper.state('isLoading')).toEqual(true)
    })
  })

  describe('findTranslation', () => {
    it('should return an error if translation is not found', () => {
      let props = {translations: [
        {
          translatedText:"Bonjour.",
          detectedSourceLanguage:"en",
          id:"kVqASjuyE",
          original:"hello."
        }
      ]
    }

      wrapper = shallow(
        <App {...props} />
      )
      let match = {params: {id: 14}, isExact: true, path: "", url: ""}
      let result = wrapper.instance().findTranslation({match})
      expect(result).toEqual('404 no translation found!')
    })

    it('should return a Detail component if a translation is clicked', () => {
      let props = {
        translations: [{translatedText: 'bonjour', id: '14'}, {translatedText: 'au revoir', id: '10'}]
      }

      let match = { params: { id: '14' }, isExact: true, path: '', url: '' }

      wrapper = shallow(
        <App {...props} />
      )

      let result = wrapper.instance().findTranslation({match})
      expect(result).toEqual(<Details id="14" translatedText="bonjour" />)
    })
  })

  describe('handleError', () => {
    it('should post an error in store', () => {
      let props = {
        postMessage: jest.fn()
      }

      wrapper = shallow(
        <App {...props} />
      )

      wrapper.instance().handleError('error message')
      expect(wrapper.instance().props.postMessage).toHaveBeenCalledWith('error message')
    })
  })
})

