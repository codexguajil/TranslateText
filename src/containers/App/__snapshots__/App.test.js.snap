// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />,
        <p
          className="message"
        />,
        <Route
          component={[Function]}
          exact={true}
          path="/translations"
        />,
        <Route
          path="/translations/:id"
          render={[Function]}
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "component": [Function],
          "exact": true,
          "path": "/",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "message",
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "component": [Function],
          "exact": true,
          "path": "/translations",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "path": "/translations/:id",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Route
            component={[Function]}
            exact={true}
            path="/"
          />,
          <p
            className="message"
          />,
          <Route
            component={[Function]}
            exact={true}
            path="/translations"
          />,
          <Route
            path="/translations/:id"
            render={[Function]}
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "message",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/translations",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "path": "/translations/:id",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
